core/consensus/component.go

```diff
@@ -106,20 +111,36 @@ func New(tcpNode host.Host, sender *p2p.Sender, peers []p2p.Peer, p2pKey *ecdsa.
 
 		// LogUponRule logs upon rules at debug level.
 		LogUponRule: func(ctx context.Context, _ core.Duty, _, round int64,
-			_ qbft.Msg[core.Duty, [32]byte], uponRule string,
+			_ qbft.Msg[core.Duty, [32]byte], uponRule qbft.UponRule,
 		) {
-			log.Debug(ctx, "QBFT upon rule triggered", z.Str("rule", uponRule), z.I64("round", round))
+			log.Debug(ctx, "QBFT upon rule triggered", z.Any("rule", uponRule), z.I64("round", round))
 		},
 
-		// LogRoundTimeout logs round timeouts at debug level.
-		LogRoundTimeout: func(ctx context.Context, duty core.Duty, process,
-			round int64, msgs []qbft.Msg[core.Duty, [32]byte],
+		// LogRoundChange logs round changes at debug level.
+		LogRoundChange: func(ctx context.Context, duty core.Duty, process,
+			round, newRound int64, uponRule qbft.UponRule, msgs []qbft.Msg[core.Duty, [32]byte],
 		) {
-			leader := leader(duty, round, peers)
-			reason := timeoutReason(round, msgs, peers, quorom, int(leader))
-			log.Debug(ctx, "QBFT round timeout",
+			fields := []z.Field{
+				z.Any("rule", uponRule),
 				z.I64("round", round),
-				z.Str("timeout_reason", reason),
+				z.I64("new_round", newRound),
+			}
+
+			steps := groupRoundMessages(msgs, len(peers), round, int(leader(duty, round, peers)))
+			for _, step := range steps {
+				fields = append(fields, z.Str(step.Type.String(), fmtStepPeers(step)))
+			}
+			if uponRule == qbft.UponRoundTimeout {
+				fields = append(fields, z.Str("timeout_reason", timeoutReason(steps, round, quorom)))
+			}
+
+			log.Debug(ctx, "QBFT round changed", fields...)
```

Comment:
add duty field in this log, it will help in debugging if there are mulitple QBFT instances running

Reply:
ctx already contains a duty in production so it would be there in prod

Reply:
in general you never have to add duty to the logs, since it should already be there